src/programmatic_policy_learning/approaches/residual_approach.py

```diff
@@ -0,0 +1,213 @@
+"""Residual policy wrapper and SB3 backend for learning action residuals."""
+
+from __future__ import annotations
+
+import contextlib
+import io
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Literal,
+    SupportsFloat,
+    TypeVar,
+    cast,
+)
+
+import gymnasium as gym
+import numpy as np
+from gymnasium import ActionWrapper
+from gymnasium.spaces import Box, Space
+from stable_baselines3 import DDPG, TD3
+from stable_baselines3.common.noise import NormalActionNoise
+from stable_baselines3.common.vec_env import DummyVecEnv
+
+from programmatic_policy_learning.approaches.base_approach import BaseApproach
+
+_ObsType = TypeVar("_ObsType")
+_ActType = TypeVar("_ActType")
+
+
+def _adapt_approach(
```

Comment:
this is a clever way to wire these things together! I think there is a simpler way we can do it though: factor out the policy from the `PendulumStupidAlgorithm` and make it just a regular python callable (like the `f` that you're creating here) and then use that instead. let me know if that makes sense -- I can also describe it in more detail.